ocaml/xapi/extauth_plugin_ADwinbind.ml

```diff
@@ -1307,6 +1288,87 @@ module RotateMachinePassword = struct
   let stop_rotate () = Scheduler.remove_from_queue task_name
 end
 
+module type LocalHostTag = sig
+  val local_ip : string
+end
+
+module HostsConfTagIPv4 : LocalHostTag = struct let local_ip = "127.0.0.1" end
+
+module HostsConfTagIPv6 : LocalHostTag = struct let local_ip = "::1" end
+
+module type HostsConf = sig
+  (* add the domain info into conf*)
+  val join : name:string -> domain:string -> lines:string list -> string list
+
+  (* remove the domain info from conf*)
+  val leave : name:string -> domain:string -> lines:string list -> string list
+end
+
+module HostsConfFunc (T : LocalHostTag) : HostsConf = struct
+  let sep = ' '
+
+  let sep_str = String.make 1 sep
+
+  type t = Add | Remove
+
+  let interest line = String.starts_with ~prefix:T.local_ip line
+
+  let handle op name domain line =
+    let line = String.lowercase_ascii line in
+    let name = String.lowercase_ascii name in
+    let domain = String.lowercase_ascii domain in
+    let fqdn = Printf.sprintf "%s.%s" name domain in
+    match interest line with
+    | false ->
+        line
+    | true ->
+        String.split_on_char sep line
+        |> List.filter (fun x -> x <> name && x <> fqdn)
+        |> (fun x -> match op with Add -> x @ [name; fqdn] | Remove -> x)
```

Comment:
Before we add something, should we make sure there is not previous entry for it? That is - should an addition always include a removal of any existing entry such that the new entry is the definitive one? Is that even possible?

Reply:
- Every entry should only exists once for an IP address
- The order does not really matter
- We remove the entry and add new one if necessary, to keep unify for `Add` and `Remove` operation.

Reply:
Does this still work when name or fqdn changes? So: the file contains an outdated entry; and the current name or fqdn does not match the previous one.